### PR TITLE
Add build api proto document

### PR DESCRIPTION
**Pull request by vankichi**
### Description:

I have created the document about how to build Vald API proto and use build clients.
For example, this document tries using Rust.
For investigation, I created the repository about sample code, you can try it.
https://github.com/vankichi/vald-sample-rust-client

### Environment:

- Go Version: 1.17.5
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.13.8

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [x] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/api/build_proto.md [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    ```bash
+    $ find proto/vald -type f -name "*.proto" | xargs sed -i "s/apis\/proto\/v1\///g"
+    $ find proto/vald -type f -name "*.proto" | xargs sed -i "s/github\.com\/googleapis\/googleapis\///g"
+    $ find proto/payload -type f -name "*.proto" | xargs sed -i "s/github\.com\/googleapis\/googleapis\///g"
+    $ find proto/payload -type f -name "*.proto" | xargs sed -i "s/github\.com\/envoyproxy\///g"
+    ```
+
+1.  Implement `build.rs` and Build proto
+
+    1. `builde.rs`
```
Comment (kmrmt) [inline docs/api/build_proto.md:140]: Is this build.rs?
```suggestion
    1. `build.rs`
```

**Review by kmrmt — CHANGES_REQUESTED** [review]

File: docs/api/build_proto.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
+# Building gRPC proto
+
+This page shows how to build gRPC proto file for calling API to your Vald Cluster.
+
+<div class="information">
+Vald provides [the official client libraries](../user-guides/sdks.md), if you can use one of SDKs we recommend to use it.
```
Comment (kevindiu) [inline docs/api/build_proto.md:6]: I don't know why the link cannot be shown, can you please try to fix it?
```suggestion
Vald provides [the official client libraries](../user-guides/sdks.md), if you can use one of SDKs we recommend to use it.
```

File: docs/api/README.md [hunk 3]
```diff
@@ -15,3 +15,6 @@ The APIs overview tables is here:
...
 | Object Service | Get information of vector(s) indexed into the Vald Agent Pods.                                                                                    | [Exists](../api/object.md#exists-rpc)<br>[GetObject](../api/object.md#getobject-rpc)<br>[StreamGetObject](../api/object.md#streamgetobject-rpc)                                                                                                                                                                                                                                                | [Vald Object APIs](../api/object.md) |
+
+And, Vald provides the public client libraries for some programming languages.
+If there are no client library for your favorite programming language, you can build yourself according to [build proto docs](../api/build_proto.md).
```
Comment (kevindiu) [inline docs/api/README.md:20]:
```suggestion
Vald provides the public client libraries for some programming languages.
If we do not provide the client library for your favorite programming language, please refer to [build proto docs](../api/build_proto.md) to build it on your environment.
```

File: docs/api/build_proto.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+</div>
+
+## Target proto files
+
+Vald defines proto file for each API.
+Let's to check the below table.
```
Comment (kevindiu) [inline docs/api/build_proto.md:12]:
```suggestion
Let's check the below table for the details.
```

File: docs/api/build_proto.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+|      Search      | [search.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/vald/search.proto) | [payload.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/payload/payload.proto)<br>[annotation.proto](https://github.com/googleapis/googleapis/blob/master/google/api/annotations.proto) | Search similar vectors with query                                  |
+|      Remove      | [remove.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/vald/remove.proto) | [payload.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/payload/payload.proto)<br>[annotation.proto](https://github.com/googleapis/googleapis/blob/master/google/api/annotations.proto) | Remove stored vectors from Vald Agent                              |
+|      Object      | [object.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/vald/object.proto) | [payload.proto](https://github.com/vdaas/vald/blob/master/apis/proto/v1/payload/payload.proto)<br>[annotation.proto](https://github.com/googleapis/googleapis/blob/master/google/api/annotations.proto) | Get object information of vector stored in Vald Agent              |
+
+## How to build proto
```
Comment (kevindiu) [inline docs/api/build_proto.md:23]:
```suggestion
## How to build protobuf
```

File: docs/api/README.md (hunk 3 above)
Comment (vankichi) [reply docs/api/README.md:20]: I think using `we do not provide` is a sense of discomfort :thinking:

File: docs/api/build_proto.md (hunk 2 above)
Comment (vankichi) [reply docs/api/build_proto.md:6]: I think it may be because this sentence is wrapped by `div`.

**Review by kmrmt — DISMISSED** [review]

File: docs/api/build_proto.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+               )
+               .unwrap_or_else(|e| panic!("protobuf compile error: {}", e));
+           Ok(())
+       }
+       ```
+
+    1. build proto
+
+       ```bash
+       $ cargo build
```
Comment (kevindiu) [inline docs/api/build_proto.md:172]: the following error occurred after I execute this command.
do you have any idea how to fix it?
```
> cargo build
    Updating crates.io index
   Compiling cfg-if v1.0.0
   Compiling proc-macro2 v1.0.36
   Compiling unicode-xid v0.2.2
   Compiling libc v0.2.116
......
   Compiling tower v0.4.11
   Compiling vald-grpc v0.1.0 (/Users/wdiu/go/src/github.com/vdaas/vald/vald-grpc)
   Compiling hyper v0.14.16
   Compiling hyper-timeout v0.4.1
   Compiling tonic v0.6.2
error: couldn't read src/client.rs: No such file or directory (os error 2)

error: could not compile `vald-grpc` due to previous error
```

Comment (kevindiu) [reply docs/api/build_proto.md:172]: I can build it after I commented the `bin` section in `Cargo.toml`.

File: docs/api/build_proto.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            Ok(())
+        }
+        ```
+
+1.  Running example
+
+    After creating Vald cluster, you can run an example code by following command.
+
+    ```bash
+    $ cargo run src/client.rs
```
Comment (kevindiu) [inline docs/api/build_proto.md:345]: it fails to execute `cargo run src/client.rs`, I am not sure if it is caused by this workaround above.
https://github.com/vdaas/vald/pull/1540/files#r796392775

I restore back the `bin` section in `Cargo.toml` before executing this step.
```
❯ cargo run src/client.rs
   Compiling vald-grpc v0.1.0 (/Users/wdiu/go/src/github.com/vdaas/vald/vald-grpc)
error[E0433]: failed to resolve: use of undeclared crate or module `vald_sample_rust_client`
 --> src/client.rs:9:5
  |
9 | use vald_sample_rust_client::payload::v1::insert;
  |     ^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `vald_sample_rust_client`

error[E0433]: failed to resolve: use of undeclared crate or module `vald_sample_rust_client`
  --> src/client.rs:10:5
   |
10 | use vald_sample_rust_client::payload::v1::object;
   |     ^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `vald_sample_rust_client`

error[E0433]: failed to resolve: use of undeclared crate or module `vald_sample_rust_client`
  --> src/client.rs:11:5
   |
11 | use vald_sample_rust_client::payload::v1::remove;
   |     ^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `vald_sample_rust_client`

error[E0433]: failed to resolve: use of undeclared crate or module `vald_sample_rust_client`
  --> src/client.rs:12:5
   |
12 | use vald_sample_rust_client::payload::v1::search;
   |     ^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `vald_sample_rust_client`
......
```

Comment (vankichi) [reply docs/api/build_proto.md:345]: Did you add the `src/lib/rs` and success import your own protobuf?

File: docs/api/build_proto.md (hunk 6 above)
Comment (vankichi) [reply docs/api/build_proto.md:173]: it's true.
I have updated the document.

**Review by kevindiu — APPROVED** [review]

**Review by kmrmt — APPROVED** [review]

**Pull request merged**